src/token.sw

```diff
@@ -25,3 +27,66 @@ pub fn force_transfer(amount: u64, asset_id: ContractId, contract_id: ContractId
         tr r3 r1 r2;
     }
 }
+
+/// Transfer `amount` coins of type `asset_id` to address `recipient`.
+pub fn transfer_to_output(amount: u64, asset_id: ContractId, recipient: Address) {
+    // note: if tx format changes, the magic number "48" must be changed !
+    // TransactionScript outputsCount has a 48 byte(6 words * 8 bytes) offset
+    // Transaction Script: https://github.com/FuelLabs/fuel-specs/blob/master/specs/protocol/tx_format.md#transactionscript
+    // Output types: https://github.com/FuelLabs/fuel-specs/blob/master/specs/protocol/tx_format.md#output
+    const OUTPUT_LENGTH_LOCATION = 48;
+    const OUTPUT_VARIABLE_TYPE = 4;
+
+    // get length of outputs from TransactionScript outputsCount:
+    let length: u8 = asm(outputs_length, outputs_length_ptr: OUTPUT_LENGTH_LOCATION) {
+        lw outputs_length outputs_length_ptr i0;
+        outputs_length: u8
+    };
+    // maintain a manual index as we only have `while` loops in sway atm:
+    let mut index: u8 = 0u8;
+    let mut outputIndex = 0;
+    let mut output_found = false;
+
+    // If an output of type `OutputVariable` is found, check if its `amount` is zero.
+    // As one cannot transfer zero coins to an output without a panic, a variable output with a value of zero is by definition unused.
+    while index < length {
+        let target_output_type_exists = asm(slot: index, type, target: OUTPUT_VARIABLE_TYPE, bytes: 8, res) {
+            xos type slot;
+            meq res type target bytes;
+            res: bool
+        };
+        // if an ouput is found of type `OutputVariable`:
+        if target_output_type_exists {
+            let amount_is_zero = asm(slot: index, a, amount_ptr, output, is_zero, bytes: 8) {
+                xos output slot;
+                addi amount_ptr output i64;
+                lw a amount_ptr i0;
+                meq is_zero a zero bytes;
+                is_zero: bool
+            };
+            // && if the amount is zero:
```

Comment:
The spec says that for a variable output the amount is the second field, and the first field `to` is a `byte[32]` so you should be adding only `i32` to `output` to get `amount_ptr`, as well as that type byte, so I'm thinking the offset should be `addi amount_ptr output i33`.

This time you're using `lw` which I hope will work from unaligned offsets @adlerjohn?

But then instead of `meq` you should just use `eq`, to compare values, or preferably do the comparison outside the asm block.

Reply:
I think adding `i32` should be sufficient, since `lw` would start from the leftmost byte.

Reply:
I will fix the offset to `i32` and extract the comparison out of the `asm` block.

Reply:
Hang on.  `XOS` returns the offset to the output, which has a `type` and `data`.  The `type` is a `uint8` but if everything is word aligned then it uses 8 bytes.  Then there's the variable data, `to`, `amount` and `asset_id`.  So the offset from what `XOS` returns and the amount is 8+32=40.

Reply:
Also, using `target_output_type_exists` from above and `amount_is_zero` here is a little weird -- you can just put those comparisons into the `if` condition.   `if type == OUTPUT_VARIABLE_TYPE { ... }` and `if amount == 0 { ... }`.

Reply:
re: "using target_output_type_exists from above "
yep, that was an oversight on my part. 
I've removed both redundant variables and made the check for equality in the `if` conditions directly.